source/server/http/admin.cc

```diff
@@ -320,6 +320,28 @@ Http::Code AdminImpl::handlerCpuProfiler(absl::string_view url, Http::HeaderMap&
   return Http::Code::OK;
 }
 
+Http::Code AdminImpl::handlerGracefulShutdown(absl::string_view, Http::HeaderMap&,
+                                              Buffer::Instance& response) {
+  if (graceful_shutdown_timer_ != nullptr) {
+    response.add("Graceful shutdown already in progress.\n");
+    return Http::Code::BadRequest;
+  }
+
+  const std::chrono::seconds shutdown_time(server_.options().parentShutdownTime());
+  ENVOY_LOG(info, "Starting graceful shutdown. Server will exit in {} seconds.",
+            shutdown_time.count());
+  graceful_shutdown_timer_ = server_.dispatcher().createTimer([this]() { server_.shutdown(); });
```

Comment:
It looks like if AdminImpl is destructed before the timeout is reached, the call to server_.shutdown() will crash.  Can we cancel the timer in the destructor if non-null, and null the timer here?

Or is this not necessary due to the way the server_.dispatcher() works?

Reply:
The timer is held in a unique_ptr. If it is destructed, the timer is cancelled/unscheduled.